rust/processor/src/worker.rs

```diff
@@ -238,9 +238,27 @@ impl Worker {
         // and write into a channel
         // The each item will be (chain_id, batch of transactions)
         let (tx, mut receiver) = tokio::sync::mpsc::channel::<(u64, Vec<Transaction>)>(BUFFER_SIZE);
-        // An Arc of mutex protected counter.
-        let channel_size = Arc::new(std::sync::Mutex::new(0));
-        let channel_size_clone = channel_size.clone();
+        // Stream closure sender.
+        let stream_closure_receiver = tx.clone();
+        let stream_closure_notifier_receiver = Arc::new(tokio::sync::Notify::new());
+        let stream_closure_notifier_sender = stream_closure_notifier_receiver.clone();
+        tokio::spawn(async move {
+            stream_closure_notifier_receiver.notified().await;
+            loop {
+                let channel_capacity = stream_closure_receiver.capacity();
+                info!(
+                    processor_name = processor_name,
+                    channel_size = BUFFER_SIZE - channel_capacity,
+                    "[Parser] Waiting for channel to be empty"
+                );
+                if channel_capacity == BUFFER_SIZE {
+                    break;
+                }
```

Comment:
Hmm if we're at the head this loop will stop right?

Reply:
Why don't we kick off this loop in the stream thread when we're at the end? (end_version > ending_version)

Reply:
done.